tools/generator/filters/rust.py

```diff
@@ -0,0 +1,174 @@
+    else:
+        return camel_case(value)
+
+
+@register_filter
+@contextfilter
+def rust_api_output_type(ctx, value: str, api_mod_path='') -> str:
+    basic_types = {
+        'bool': 'bool',
+        'double': 'f64',
+        'int': 'i32',
+        'string': 'String',
+        'void': '()',
+    }
+
+    if is_array(value):
+        return 'Vec<{}>'.format(rust_api_output_type(ctx, get_array_inner(value)))
+    elif value in basic_types:
+        return basic_types[value]
+    elif is_struct(ctx, value) and is_tuple(ctx['game'].get_struct(value)):
+        return '({})'.format(
+            ', '.join(
+                rust_api_output_type(ctx, field)
+                for _, field, _ in ctx['game'].get_struct(value)['str_field']
+            ))
+    else:
+        return api_mod_path + camel_case(value)
+
+
+@register_filter
+@contextfilter
+def rust_api_input_type(ctx, value: str, api_mod_path='', skip_ref=False) -> str:
+    if value == 'string':
+        return '&str'
+    elif is_array(value):
+        return '&[' + rust_api_output_type(ctx, get_array_inner(value)) + ']'
+    elif is_struct(ctx, value) and is_tuple(ctx['game'].get_struct(value)):
+        return '({})'.format(
+            ', '.join(
+                rust_api_input_type(ctx, field)
+                for _, field, _ in ctx['game'].get_struct(value)['str_field']
+            ))
+    elif not rust_is_copy(ctx, value):
+        if skip_ref:
+            return rust_api_output_type(ctx, value, api_mod_path)
+        else:
+            return '&' + rust_api_output_type(ctx, value, api_mod_path)
+    else:
+        return rust_api_output_type(ctx, value, api_mod_path)
+
+
+@register_filter
+@contextfilter
+def rust_is_copy(ctx, value: str) -> bool:
+    """Check if a type implements the Copy trait"""
+    if is_struct(ctx, value) and is_tuple(ctx['game'].get_struct(value)):
+        return all(
+            rust_is_copy(ctx, field_type)
+            for _, field_type, _ in ctx['game'].get_struct(value)['str_field']
+        )
+
+    return (
+        is_enum(ctx, value) or value in ['bool', 'double', 'int', 'void']
+    )
+
+
+@register_filter
+@contextfilter
+def rust_auto_traits(ctx, value: str) -> set:
+    """
+    Return the list of auto traits that can be implemented for given input
+    type.
+    """
+    if is_struct(ctx, value):
+        inherited = set.intersection(*(
+            rust_auto_traits(ctx, field_type)
+            for _, field_type, _ in ctx['game'].get_struct(value)['str_field']
+        ))
+
+        if not is_tuple(ctx['game'].get_struct(value)) and 'Copy' in inherited:
+            inherited.remove('Copy')
+
+        return inherited
+
+    if is_array(value):
+        inherited = rust_auto_traits(ctx, get_array_inner(value))
+
+        if 'Copy' in inherited:
+            inherited.remove('Copy')
+
+        return inherited
+
+    if value == 'double':
+        return {'Copy', 'Clone', 'Debug', 'PartialEq', 'PartialOrd'}
```

Comment:
Returning a set here might get you randomly ordered traits, which isn't great for gendiff etc. Why don't you return a list instead?

Reply:
That's because I am computing intersections over this result, that's why I am using the sort filter in the template. Is this still a big issue?

Reply:
That's because I am computing intersections over this result, that's why I am using the sort filter in the template. Is this still a big issue?

Reply:
Ah, I hadn't seen the |sort filter, looks good then